core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,839 @@
+
+  /**
+   * The number of elements in the chunk.
+   */
+  def length: Int
+
+  /**
+   * Determines if the chunk is empty.
+   */
+  final def isEmpty: Boolean = length == 0
+
+  /**
+   * Determines if the chunk is not empty.
+   */
+  final def notEmpty: Boolean = length > 0
+
+  /**
+   * Returns the concatenation of this chunk with the specified chunk.
+   */
+  final def ++[A1 >: A](that: Chunk[A1]): Chunk[A1] =
+    if (self.length == 0) that
+    else if (that.length == 0) self
+    else Chunk.Concat(self, that)
+
+  /**
+   * Returns a filtered, mapped subset of the elements of this chunk.
+   */
+  def collect[@specialized B](p: PartialFunction[A, B]): Chunk[B] =
+    self.materialize.collect(p)
+
+  /**
+   * Drops the first `n` elements of the chunk.
+   */
+  final def drop(n: Int): Chunk[A] =
+    self match {
+      case Chunk.Slice(c, o, l) =>
+        Chunk.Slice(c, o + n, l - n)
+      case c @ Chunk.Singleton(_) =>
+        if (n <= 0) c else Chunk.empty
+      case _ =>
+        Chunk.Slice(self, n, self.length - n)
+    }
+
+  /**
+   * Drops all elements so long as the predicate returns true.
+   */
+  def dropWhile(f: A => Boolean): Chunk[A] = {
+    val len = self.length
+
+    var i = 0
+    while (f(self(i)) && i < len) {
+      i += 1
+    }
+
+    if (i == len) Chunk.empty
+    else Chunk.Slice(self, i, len - i)
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Chunk[_] =>
+      if (self.length != that.length) false
+      else {
+        var i     = 0
+        var equal = true
+        val len   = self.length
+
+        while (equal && i < len) {
+          equal = self(i) == that(i)
+          i += 1
+        }
+
+        equal
+      }
+    case _ => false
+  }
+
+  /**
+   * Returns a filtered subset of this chunk.
+   */
+  def filter(f: A => Boolean): Chunk[A] = {
+    implicit val B = Chunk.classTagOf(this)
+
+    val len  = self.length
+    val dest = Array.ofDim[A](len)
+
+    var i = 0
+    var j = 0
+    while (i < len) {
+      val elem = self(i)
+
+      if (f(elem)) {
+        dest(j) = elem
+        j += 1
+      }
+
+      i += 1
+    }
+
+    if (dest == null) Chunk.Empty
+    else Chunk.Slice(Chunk.Arr(dest), 0, j)
+  }
+
+  /**
+   * Returns the concatenation of mapping every element into a new chunk using
+   * the specified function.
+   */
+  def flatMap[@specialized B](f: A => Chunk[B]): Chunk[B] = {
+    val len                    = self.length
+    var chunks: List[Chunk[B]] = Nil
+
+    var i               = 0
+    var total           = 0
+    var B0: ClassTag[B] = null.asInstanceOf[ClassTag[B]]
+    while (i < len) {
+      val chunk = f(self(i))
+      if (chunk.length > 0) B0 = Chunk.classTagOf(chunk)
+      chunks ::= chunk
+      i += 1
+      total += chunk.length
+    }
+    chunks = chunks.reverse
+
+    if (B0 == null) Chunk.empty
+    else {
+      implicit val B = B0
+
+      val dest: Array[B] = Array.ofDim(total)
```

Comment:
just wondering if it would be faster to just make an initial guess, and copy the elements to an exponentially growing array (using array copies, since they can be linearised very well by the CPU).

Reply:
Well, we do have the length of all the intermediate chunks at hand - it is computed anyway when constructing a chunk (`Chunk.length` is always a constant time operation). So `total` is not a guess and should be the accurate size.

Reply:
Oh, sorry, length is actually not a constant time operation. Specialization issues made that impossible.

In that case it could be faster but should probably be benchmarked. We can do this in a subsequent issue - care to open one?

Reply:
no, never mind, I misread the code